backend/svirt.pm

```diff
@@ -270,35 +296,74 @@ sub start_serial_grab {
 
-    bmwqemu::diag("Starting SSH connection to connect to libvirt domain $name via serial port $port");
-    my $credentials = $self->read_credentials_from_virsh_variables;
-    my $ssh         = $self->new_ssh_connection(%$credentials);
-    my $chan        = $ssh->channel();
-    die 'No channel found' unless $chan;
-    $chan->blocking(0);
-    $chan->pty('vt100', {echo => 1});
-    $chan->pty_size(1024, 24);
-    $chan->shell();
-    print($chan "PS1='# '\n");
+Opens SSH connection to grab the serial console for kernel log (serial0.txt)
+or serial terminal log (serial_terminal.txt).
+
+C<$args{port}> used non-default port (used for serial_terminal.txt)
+C<$args{devname}> used device name (usually console, for serial_terminal.txt)
+C<$args{is_terminal}> for serial terminal (usually console, for serial_terminal.txt)
+=cut
+sub open_serial_console_via_ssh {
+    my ($self, $name, %args) = @_;
+    my ($chan, $cmd, $cmd_full, $ret, $ssh, $stderr, $stdout);
+    my $port    = $args{port}    // '';
+    my $devname = $args{devname} // '';
+    my $marker  = "CONSOLE_EXIT_" . get_required_var('JOBTOKEN') . ":";
+    my $log     = $self->serial_terminal_log_file();
 
-    # note: see comments in start_serial_grab for the special handling of vmware/hyperv
     if (check_var('VIRSH_VMM_FAMILY', 'vmware')) {
-        $chan->exec('nc ' . get_var('VMWARE_SERVER') . ' ' . $port);
+        # libvirt esx driver does not support `virsh console', so
+        # we have to connect to VM's serial port via TCP which is
+        # provided by ESXi server.
+        $cmd = 'nc ' . get_var('VMWARE_HOST') . ' ' . $port;
     }
     elsif (check_var('VIRSH_VMM_FAMILY', 'hyperv')) {
-        $chan->exec('nc ' . get_var('HYPERV_SERVER') . ' ' . $port);
+        # Hyper-V does not support serial console export via TCP, just
+        # windows named pipes (e.g. \\.\pipe\mypipe). Such a named pipe
+        # has to be enabled by a namedpipe-to-TCP on HYPERV_SERVER application.
+        $cmd = 'nc ' . get_var('HYPERV_SERVER') . ' ' . $port;
     }
     else {
-        $chan->exec("virsh console \"$name\" \"serial$port\"");
+        $cmd = "virsh console $name $devname$port";
+    }
+
+    $cmd_full = "script -f $log -c '$cmd; echo \"$marker \$?\"'";
+    bmwqemu::diag("Starting SSH connection to connect to libvirt domain '$name' (cmd: '$cmd'), full cmd: '$cmd_full'");
+
+    ($ssh, $chan) = $self->run_ssh_cmd($cmd_full, nonblock => 1);
+    ($ret, $stdout, $stderr) = $self->run_ssh_cmd("grep -q '^$marker' $log");
+    if (!$ret) {
+        (undef, $stdout, undef) = $self->run_ssh_cmd("cat $log");
+        $self->die("problem with virsh: cmd: '$cmd', output of script wrapper: '$stdout')");
     }
 
     return ($ssh, $chan);
 }
 
+sub delete_log
+{
+    my ($self) = @_;
+    my $log = $self->serial_terminal_log_file();
+    $self->run_ssh_cmd("[ -f '\$log' ] && rm -v \$log; done");
+}
+
+# TODO: use CORE::GLOBAL::die
```

Comment:
Do you really intend to do this? If not then just comment that it may be necessary to change this to CORE::GLOBAL::die.

Reply:
@richiejp make sense, changed both.

@okurz here is a fix for package missing `IO::Scalar ` dependency:
https://build.opensuse.org/request/show/684257

Reply:
This cannot be merged until @mnowaksuse tests/approves it.